text/0028-scaling-codeowners.md

```diff
@@ -0,0 +1,63 @@
+
+### Code Owners Increases Auto Assignment
+
+As we make efforts to increase the signal-to-noise ratio in our notifications and Issue stream - Being able to answer “Who is this Issue important for?” is a necessary opinion that Sentry needs to build. Code Owners enables auto assignment of Issues which is one of the clearest signals of issue ownership we have today. We want more auto assignment of Issues in Sentry and thus we need more Orgs to use Code Owners
+
+# Goals:
+
+1. Users can upload a Code Owners File Size that's up to 3Mb in size
+2. The processing time to calculate assignments should not increase significantly. 
+
+# How it Works (simplified)
+
+We have Code Mappings that map the transformation of the source code to the production build. Code Mappings help us reverse-engineer the file paths from an error’s stacktrace to source code file paths. 
+
+A customer can apply a Code Mapping to a CODEOWNERS file to get the file paths that should match against the stacktrace file paths. The transformed output is stored in `sentry_projectcodeowners` and is what Sentry will use to determine assignments.
+
+When Sentry ingests a new event, we go through each line in our transformed CODEOWNERS and see if each line has a match across any of the frames in the stacktrace. We only use the last matching line for the assignment.
+
+# Current System Constraints
+
+- We store the original and transformed CODEOWNERS file in a `sentry_projectcodeowners` row. However to prevent overloading Postgres from queries on every event, we cache the row in Memcache. Memcache can only support 1mb. Rows larger than 1mb will hit the database directly. Since we store the original and transformed CODEOWNERS in the same row, files larger than 500k chars will exceed this limitation.
+- On each event, the assignment calculation is currently O((m*l + o)*n) time complexity, where *m* is the # of lines in the CODEOWNERS file, *l* is the number of code mappings mapped to the project, *o* is the number of lines in the Ownership Rules and n is the number of lines in the stacktrace. Essentially quadratic.
+- We recalculate the assignment for each event, even if the Ownership Rules and CODEOWNERS file has not been updated.
+
+# Proposal
+
+First we need baseline metrics. We need to know how long it currently takes a worker to calculate the assignments with a 100k file.
+
+> Memcache can only support 1mb.
+
+To handle Memcache’s 1mb limit, we can zip to compress the rows and store that. However, we need to consider the memory limits of our workers during the unzip process. We need to consider the impact of (3mb CODEOWNERS raw file  + 3mb CODEOWNERS transformed file)* x number of code mappings applied to that file for a project is potentially unbounded. Two options:
```

Comment:
Here are some changes I suggest:
* Do not store the untransformed codeowners
* Do not store a row per code mapping but a row per repo
  * All code mappings for a specific repo should be applied at the same time (e.g. sentry/ and sentry_plugins/ for the Sentry repo)

This means we would have:
```diff
- 3mb CODEOWNERS raw file  + 3mb CODEOWNERS transformed file)* x number of code mappings
+                            3mb CODEOWNERS transformed file)* x number of repos with a code mapping
```

Also some other representation changes:
* Do not store comments
* Create a data structure that groups paths into buckets of teams
  * e.g. team A matches list of paths
  * Perhaps this should lead to a new table to store a more optimized structured to replace the current one
  * I would consider moving away completely from storing things raw but create a model(s) that could represent the information you care about
* Instead of storing team as strings, we should store them a ID to real teams. If we don't have team matches we don't have any specific team to match but everyone in the project

When reading the data from either the DB or memcache you can construct a more optimized structure based on what is stored.

For instance, you may want to store a tree where each node is part of the path while the leaves are either an asterisk or file name.

Reply:
I suspect that loading a 3mb file from either the database or memcache will be pretty memory intensive as well. I think we should probably do some kind of profiling on how long it takes to fetch even a compressed file from memcache